### ISSUE TITLE
Move Twitch plugin to use Twitch API v3

### ISSUE THREAD

**Issue by thinkpad4** [streamlink/streamlink#20]
Or merge javiercantero's v3-api branch https://github.com/javiercantero/livestreamer/commits/v3-api

**Comment by javiercantero** [comment]
Note that some of the API used by livestreamer to get the stream is not documented and apparently not public, so it can change without prior warning, be deprecated or not comply with the rest of the specifications. For example, the API call efbc85ea613fd0a200ca65a6f6320be0c775b771/src/livestreamer/plugins/twitch.py#L186 which doesn't work with the kraken API format (api.twitch.tv/kraken/). So I'm a little afraid of changing anything that it's not documented even when I've checked that the change works right now.

**Comment by thinkpad4** [comment]
Jariercantero, well I'm using your version of twitch.py and it is working perfect, I would chance to say it seems to load the stream faster then before

**Comment by cdrage** [comment]
Any reason to use the undocumented API v3 *yet*? Hold off until it's 100% documented and available.

**Comment by javiercantero** [comment]
@cdrage The API v3 is totally documented and working (see [here](https://github.com/justintv/Twitch-API)). What is not documented (with any API version) or **supported** is to [download Twitch content outside their player](https://www.reddit.com/r/Twitch/comments/2p0zka/tardsplaya_stopped_working_for_everyone/cmsjqgj)). livestreamer and other clients work because they use the same calls that their client uses to get the stream, but they could modify it at will.

**Comment by bastimeyer** [comment]
> undocumented API v3

You're mixing things up here a bit... Both public and private Twitch APIs are being used.

There are two supported versions of the public `/kraken` namespace right now, namely v2 and v3. When no API version is specified in the request, the latest one is always assumed. Setting an explicit version header will ensure that nothing will break once a newer version gets released. See https://github.com/justintv/Twitch-API

The private namespace that's being used is `/api` and also `tmi.twitch.tv` for checking for hosted channels. See https://github.com/streamlink/streamlink/blob/ef3476952fa9e6fa66f50e0fa515cabe070f7360/src/streamlink/plugins/twitch.py#L193-L218

I think there are a couple of API calls that could be changed so that the public namespace is used instead of the private one. I might be wrong, though.

**Comment by javiercantero** [comment]
> I think there are a couple of API calls that could be changed so that the public namespace is used instead of the private one. I might be wrong, though.

No, you are not: `/api/viewer/*` also works as `/kraken/viewer/*` (but it's not part of the public API)

**Comment by bastimeyer** [comment]
I have just seen that this issue has not been resolved yet...

Twitch has added a new API version (v5) two months ago in mid november and announced that v1 and v2 will be removed **2017-02-14**. As I've said in my post above, not specifying an API version defaults to the latest version (which is v3), but this will change when v1/v2 have been removed. v5 will then become the default. There's also a deadline for v3, **2018-02-14**.

> ACTION REQUIRED: Twitch API Version Update
> TL;DR: Read all of this. It’s important and will break your integrations. We’re changing how the Twitch API works and old versions of our API will be deprecated and removed.
> 
> Today, we’re delivering a change that you’ve requested for quite some time! Twitch API v5 is available now and uses IDs to reference users instead of usernames. This new API version puts us in a better position to add new functionality and features to both the API and the Twitch website.
> 
> You can start using v5 and upgrading your integrations today. You simply pass the application/vnd.twitchtv.v5+json header on your requests. We’re working hard to deliver reference documentation on the affected endpoints by mid-December. Here is a quick example:
> 
> Retrieving a user in v3
> curl -H 'Client-ID: XXXXX' \
> -H 'Accept: application/vnd.twitchtv.v3+json' \
> https://api.twitch.tv/kraken/users/dallas
> 
> Retrieving a user in v5
> curl -H 'Client-ID: XXXXX' \
> -H 'Accept: application/vnd.twitchtv.v5+json' \
> https://api.twitch.tv/kraken/users/44322889
> 
> There are other changes to note with this version:
> The _links object is removed from all responses. As a developer, you will be responsible for handling the paging offsets when polling the API.
> Added sub-second precision to time fields.
> Removed /games/featured endpoint.
> Removed delay, background, and banner from the channels response object.
> Replaced q and type parameters on the search endpoint. The q parameter is now called query. The type parameter is always set to suggest.
> Removed paywalled field from the video response object.
> Removed /kraken/chat, /kraken/chat/emoticons, and /kraken/chat/emoticon_index.
> 
> What does this mean for me as a developer?
> You need to migrate your code to use v5. On February 14, 2017, we will be removing v1 and v2 of the API and transition v3 into a deprecated state. In one year (February 14, 2018), v3 will be removed from the API. We hope that this gives enough transition time for your integrations.
> If your integration relies on username now, you can use the /users endpoint to translate from a username to an ID as shown below.
> 
> curl -H 'Client-ID: XXXXX' \
> -H 'Accept: application/vnd.twitchtv.v3+json' \
> https://api.twitch.tv/kraken/users/dallas
> For questions regarding this transition, please ask on the developer forum thread. We know this is a big transition and want to help in any way that we can.
> 
> Thank you for being a part of the Twitch Developer community!

**Comment by javiercantero** [comment]
I haven't read the v5 API yet, so I'm going to push some initial patches to upgrade to v3.

**Comment by javiercantero** [comment]
@thinkpad4 Could you change the title to "Move Twitch plugin to use Twitch API v5"? Thanks.

**Comment by beardypig** [comment]
Maybe it's better to open a new issue to track the move to v5 and reference this one?

**Comment by javiercantero** [comment]
@beardypig Or that and close this. Whatever you all prefer.

**Comment by gravyboat** [comment]
@javiercantero Let's close this one out and create a new issue to move to v5.
